libs/core/async_mpi/tests/unit/algorithm_transform_mpi.cpp

```diff
@@ -0,0 +1,246 @@
+        }
+
+        // Operator| overload
+        {
+            // MPI function pointer
+            int data = 0, count = 1;
+            if (rank == 0)
+            {
+                data = 42;
+            }
+            auto result = ex::just(&data, count, datatype, 0, comm) |
+                mpi::transform_mpi(MPI_Ibcast) | ex::sync_wait();
+            if (rank != 0)
+            {
+                HPX_TEST_EQ(data, 42);
+            }
+            if (rank == 0)
+            {
+                HPX_TEST(result == MPI_SUCCESS);
+            }
+        }
+
+        // Failure path
+        {
+            // Exception with error sender
+            bool exception_thrown = false;
+            try
+            {
+                mpi::transform_mpi(error_sender<int*, int, MPI_Datatype, int,
+                        MPI_Comm>{}, MPI_Ibcast) | ex::sync_wait();
+                HPX_TEST(false);
+            }
+            catch (std::runtime_error const& e)
+            {
+                HPX_TEST_EQ(std::string(e.what()), std::string("error"));
+                exception_thrown = true;
+            }
+            HPX_TEST(exception_thrown);
+        }
+
+        {
+            // Exception in the lambda
+            bool exception_thrown = false;
+            int data = 0, count = 1;
+            auto s = mpi::transform_mpi(ex::just(&data, count, datatype, 0, comm),
+                    [](int *data, int count, MPI_Datatype datatype, int i,
+                        MPI_Comm comm, MPI_Request* request)
+                    {
+                        MPI_Ibcast(data, count, datatype, i, comm, request);
+                        throw std::runtime_error("error in lambda");
+                    });
+            try {
+                ex::sync_wait(s);
+            }
+            catch (std::runtime_error const& e)
+            {
+                HPX_TEST_EQ(std::string(e.what()), std::string("error in lambda"));
+                exception_thrown = true;
+            }
+            HPX_TEST(exception_thrown);
+        }
+        // Necessary to avoid a seg fault caused by MPI data going out of scope
+        // too early when an exception occured outside of MPI
+        MPI_Barrier(comm);
+
+        {
+            // Exception thrown through HPX custom error handler that throws
+            int* data = nullptr, count = 0;
+            bool exception_thrown = false;
+            try
+            {
+                mpi::transform_mpi(ex::just(data, count, datatype, -1, comm), MPI_Ibcast) | ex::sync_wait();
+                HPX_TEST(false);
+            }
+            catch (std::runtime_error const& e)
+            {
+                HPX_TEST(std::string(e.what()).find(std::string("invalid root"))
+                        != std::string::npos);
+                exception_thrown = true;
+            }
+            HPX_TEST(exception_thrown);
+        }
+        // let the user polling go out of scope
+        }
```

Comment:
Yes I was planning of doing a clang-format commit when the PR is ready as otherwise it can mess up with my commit fixups, will be solved when there is no other review comments :)